### PR TITLE
Custom Trainer subclass for quantization and train script

### PR DESCRIPTION
**Pull request by Clydingus**
Yeah lots of stuff here

TODO:

1. GPTQConfig quant still not working. Have to figure out a _clean_ way to turn eval_dataloader into a list and make this list consistent globally. 
2. train.py script still has undone sections, check the comments
3. update the .yaml config files to include some new stuff
4. decide what lr scheduler to use, currently its OneCycleLR but we might want to switch to CosineAnnealing
5. @photomz double check some changes I made
6. unit tests for the changes I made to dataloader logic as it was bugged wrt. how it padded inputs
7. test_modeling_softerllama.py has some commented out portions as im testing trainer evaluation loop using that. Will uncomment after.
8. Test trainer on softerbert as well by including a similar trainer evaluation unit test in test_modeling_softerbert.py
9. ????
10. Profit.

### CODE REVIEW

File: src/dataloader.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+            k: [t[i : i + self.max_seq_length] for i in range(0, total_length, self.max_seq_length)]
+            for k, t in concatenated_examples.items()
+        }
+        return result
+
+    def encode(self, example_batch):
+        # tokenize the text
+        features = self.tokenizer(
+            example_batch["text"],
+            # TODO: this line is trying to pad but there is no guarantee that we actually have a PAD token instantiated
```
Comment (Clydingus) [inline src/dataloader.py]: @photomz Yeah this is the padding issue I mentioned. We have to create a unit test for this per your comments. Just double check in the PR. I'll write the unit test later.

File: src/softerllama/modeling_softerllama.py [hunk 2]
```diff
@@ -195,7 +195,7 @@ def forward(
         # upcast attention to fp32
         # softermax overriding original attention softmax
         attn_weights = softermax(attn_weights, self.n_bias, dim=-1).type(dtype=torch.float32).to(query_states.dtype)
-        attn_weights = clip_logits(attn_weights, self.n_clip, dim=-1)
+        attn_weights = clip_logits(attn_weights, self.n_clip)
```
Comment (Clydingus) [inline src/softerllama/modeling_softerllama.py:198]: @photomz double check clip_logits was throwing an error since `dim=-1` doesn't exist? Can you check that deleting it is fine and I didn't have to transpose something.

File: src/quantization/quant.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+default quantization config settings and a function that will quantize a given model.
+"""
+from transformers import PreTrainedModel, GPTQConfig
+from torch.utils.data import DataLoader
+
+# default quantization configuration, requires the dataset be specified
+default_quant_configs = {"bits": 4, "group_size": 128, "desc_act": False}
+
+
+def quantize(model: PreTrainedModel, quant_config: GPTQConfig):
```
Comment (photomz) [inline src/quantization/quant.py]: @Clyde013 does this quant code work?

File: src/training_utils.py [hunk 4]
```diff
@@ -20,27 +40,47 @@ def evaluate(
...
+        print("regular model eval")
+        orig_metrics = super().evaluate(eval_dataset, ignore_keys, metric_key_prefix="orig")
+
+        # store a ref to the original model while moving it to cpu
+        orig_model = self.model.to("cpu")
+        # quantize the model
+        print("quantizing")
+        logger.info("Quantizing model for eval loop")
+        
+        quant_config = GPTQConfig(dataset=self.eval_dataset, **self.args.quant_kwargs)
```
Comment (photomz) [inline src/training_utils.py]: Ignore if GPT-Q quant isn't working for now?

File: train.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+model_mapping = {"softerllama": SofterLlamaForCausalLM, "softerbert": SofterBertForMaskedLM}
+tokenizer_mapping = {"softerllama": LlamaTokenizerFast, "softerbert": BertTokenizer}
+
+# the config files are loaded automatically into wandb.config using wandb --configs command line argument
+wandb.init(project="training-runs", entity="softermax")
+wandb.run.name = wandb.config.run_name
+wandb.run.save()
+
+# model configs setup
+# TODO
```
Comment (photomz) [inline train.py]: Not complete?

**Review by photomz — COMMENTED** [review]
PR seems incomplete
